### PR TITLE
No panic on timestamp buffer overflow

### LINKED ISSUE #3512: Overflow when casting timestamps prior to the epoch
**Describe the bug**
When a string such as `'1677-06-14T07:29:01.256'` is casted to a timestamp (via `string_to_timestamp_nanos`) it panics in debug builds and gives a wonky answer in release builds. 

**To Reproduce**
Run this test in `arrow-cast/src/display.rs`

```rust
    #[test]
    fn string_to_timestamp_old() {
        assert_eq!(
            123,
            parse_timestamp("1677-06-14T07:29:01.256").unwrap()
        );
    }
```
```

failures:

---- parse::tests::string_to_timestamp_old stdout ----
thread 'parse::tests::string_to_timestamp_old' panicked at 'attempt to multiply with overflow', /Users/alamb/.cargo/registry/src/github.com-1ecc6299db9ec823/chrono-0.4.23/src/naive/datetime/mod.rs:426:21
stack backtrace:
   0: rust_begin_unwind
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:65:14
   2: core::panicking::panic
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:115:5
   3: chrono::naive::datetime::NaiveDateTime::timestamp_nanos
             at /Users/alamb/.cargo/registry/src/github.com-1ecc6299db9ec823/chrono-0.4.23/src/naive/datetime/mod.rs:426:21
   4: arrow_cast::parse::string_to_timestamp_nanos
             at ./src/parse.rs:102:19
   5: arrow_cast::parse::tests::parse_timestamp
             at ./src/parse.rs:544:22
   6: arrow_cast::parse::tests::string_to_timestamp_old
             at ./src/parse.rs:471:13
   7: arrow_cast::parse::tests::string_to_timestamp_old::{{closure}}
             at ./src/parse.rs:468:5
   8: core::ops::function::FnOnce::call_once
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/ops/function.rs:251:5
   9: core::ops::function::FnOnce::call_once
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/ops/function.rs:251:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    parse::tests::string_to_timestamp_old
```

**Expected behavior**
It should not panic; It should return an error

**Additional context**
Found in datafusion https://github.com/apache/arrow-datafusion/issues/4875 by @DDtKey 

The actual panic is in chrono, it appears to be this line

https://github.com/apache/arrow-rs/blob/55c87c114443739ed73ebc28d0ba53bf875ecd9a/arrow-cast/src/parse.rs#L99-L103

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #3512.

# Rationale for this change
No panic on timestamp buffer overflow, when casting happens for nanoseconds for dates beyond supported by `chrono-rs`

# What changes are included in this PR?
Defensive check added to prevent panic

# Are there any user-facing changes?
No

### CODE REVIEW

**Review by tustvold — APPROVED** [review]
It might be nice to include a test within the cast kernel or something

**Comment by comphead** [comment]
> It might be nice to include a test within the cast kernel or something

Thanks @tustvold for your feedback, there is already a test attached to test old date. 
https://github.com/apache/arrow-rs/pull/3519/files#diff-cd568dd408cf11db340aa0259545de48fca7a0c0ac680bb9ff9a34956bab52b1R864

Please let me know if that is not enought

**Review by viirya — APPROVED** [review]

File: arrow-cast/src/parse.rs [hunk 1]
```diff
@@ -373,6 +385,9 @@ impl Parser for Time32SecondType {
 /// Number of days between 0001-01-01 and 1970-01-01
 const EPOCH_DAYS_FROM_CE: i32 = 719_163;
 
+/// Error message if nanosecond conversion request beyond supported interval
+const ERR_NANOSECONDS_NOT_SUPPORTED: &str = "The dates that can be represented as nanoseconds have to be between 1677-09-21T00:12:44.0 and 2262-04-11T23:47:16.854775804";
```
Comment (alamb) [inline arrow-cast/src/parse.rs:389]: ❤️

**Review by alamb — COMMENTED** [review]
Thank you @comphead !

**Pull request merged**